audmath/core/api.py

```diff
@@ -86,6 +87,220 @@ def db(
+        0.002
+        >>> duration_in_seconds('ms')
+        0.001
+        >>> duration_in_seconds('2000', 1000)
+        2.0
+        >>> duration_in_seconds(np.timedelta64(2, 's'))
+        2.0
+        >>> duration_in_seconds(pd.to_timedelta(2, 's'))
+        2.0
+
+    """  # noqa: E501
+    # Dictionary with allowed unit entries
+    # and mapping from pandas.to_timedelta()
+    # to numpy.timedelta64, see
+    # https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units
+    unit_mapping = {
+        # week
+        'W': 'W',
+        # day
+        'D': 'D',
+        'days': 'D',
+        'day': 'D',
+        # hour
+        'h': 'h',
+        'hours': 'h',
+        'hour': 'h',
+        'hr': 'h',
+        # minute
+        'm': 'm',
+        'minutes': 'm',
+        'minute': 'm',
+        'min': 'm',
+        'T': 'm',
+        # second
+        's': 's',
+        'seconds': 's',
+        'second': 's',
+        'sec': 's',
+        'S': 's',
+        # millisecond
+        'ms': 'ms',
+        'milliseconds': 'ms',
+        'millisecond': 'ms',
+        'millis': 'ms',
+        'milli': 'ms',
+        'L': 'ms',
+        # microsecond
+        'us': 'us',
+        'μs': 'us',
+        'microseconds': 'us',
+        'microsecond': 'us',
+        'micros': 'us',
+        'micro': 'us',
+        'U': 'us',
+        # nanosecond
+        'ns': 'ns',
+        'nanoseconds': 'ns',
+        'nanosecond': 'ns',
+        'nanos': 'ns',
+        'nano': 'ns',
+        'N': 'ns',
+    }
+
+    # numpy.timedelta64() accepts only integer as input values,
+    # so we need to convert all values
+    # to nanoseconds and integers first
+    def to_nanos(value, unit):
+        if unit == 'W':
+            value = value * 7 * 24 * 60 * 60 * 10 ** 9
+        elif unit == 'D':
+            value = value * 24 * 60 * 60 * 10 ** 9
+        elif unit == 'h':
+            value = value * 60 * 60 * 10 ** 9
+        elif unit == 'm':
+            value = value * 60 * 10 ** 9
+        elif unit == 's':
+            value = value * 10 ** 9
+        elif unit == 'ms':
+            value = value * 10 ** 6
+        elif unit == 'us':
+            value = value * 10 ** 3
+        return int(value)
+
+    if isinstance(duration, str):
+
+        # ensure we have a str and not numpy.str_
+        duration = str(duration).strip()
+
+        if not duration:
+            raise ValueError(
+                "'duration' is not allowed to be empty, "
+                "or contain only spaces."
+            )
+
+        spaces = duration.count(' ')
+        if spaces > 1:
+            raise ValueError(
+                "You must only include one space (' ') "
+                "between the value and the unit. "
+                f"Your string '{duration}' contains {spaces}."
+            )
+        elif spaces == 1:
+            value, unit = duration.split(' ')
+        else:
+            match = re.match('^([0-9]*[.]?[0-9]*)([a-zA-Zμ]*)$', duration)
```

Comment:
We could expand the pattern to match also a single space between the value and the unit with:

```
'^([0-9]*[.]?[0-9]*) ?([a-zA-Zμ]*)$'
```

then we wouldn't need all the code above and could start with match after the
```python
duration = str(duration).strip()
```
line.

The only downside would be that we loose the two custom error messages regarding an empty string or too many spaces.

Any preference?

Reply:
Hehe, I was just about to write that I'm not convinced if we really need all those custom error messages. E.g. we now have:

```python
>>> audmath.duration_in_seconds('2.m5s')
ValueError: Your given duration '2.m5s' is not conform to the 'value' 'unit' pattern.
```

but:

```python
>>> audmath.duration_in_seconds('2.m5 s')
ValueError: could not convert string to float: '2.m5'
```

To me, only having the latter error in all cases would be sufficient.

So yes, if it is possible please go ahead with a single pattern. I would maybe slightly change it to allow an arbitrary number of spaces between the value and the unit. The case with a only a unit (e.g. 'ms') I'm not sure if we should really support, I guess it's rather a mistake if a user misses the value.

Reply:
I did now the following changes:

* I compile the search pattern only during import of the module, so we don't have to do it every time we are calling `duration_to_seconds()`
* I included the parsing of spaces in the pattern and raise now a single error message if the pattern is not matched, namely `ValueError: Your given duration '{duration}' is not conform to the 'value' 'unit' pattern.`
* I allowed for one or several spaces between value and unit and a added a test for two spaces between value and unit

In my opinion, the support for `audmath.duration_in_seconds('ms')` is a nice feature ;)

Reply:
> In my opinion, the support for audmath.duration_in_seconds('ms') is a nice feature ;)

I'm not convinced, but ok I don't mind :)

There is still one case we do not capture:

```python
>>> audmath.duration_in_seconds('1.3')
You have to provide 'sampling_rate' when specifying the duration in samples as you did with '1.3'.
```

but then:

```python
>>> audmath.duration_in_seconds('1.3', sampling_rate=1)
ValueError: invalid literal for int() with base 10: '1.3'
```

In the first case we could simply say that `'1.3'` converts to 1.3, but then it probably makes sense that `'1'` without sampling rate converts to 1.0 instead of raising an error.

Reply:
And in principle, we could also allow support samples given as a floating point number, e.g. `audmath.duration_in_seconds('0.5', sampling_rate=2)` could return 1.0.

Reply:
Makes sense, we now have:

```python
>>> audmath.duration_in_seconds('1.3')
...
ValueError: You have to provide 'sampling_rate' when specifying the duration in samples as you did with '1.3'.
>>> audmath.duration_in_seconds('1.3', sampling_rate=1)
1.0
```

Reply:
But I do not round, but use `int` at the moment, so we get:

```python
>>> audmath.duration_in_seconds('0.5', sampling_rate=2)
0.0
```
which I guess is not what you would expect?

Reply:
The problem is not that I use `int`, but that I use `int` on the sample value and do not first divide by sampling rate. Will change it.

Reply:
I fixed it, now we get:

```python
>>> audmath.duration_in_seconds('1.3', sampling_rate=1)
1.3
>>> audmath.duration_in_seconds('0.5', sampling_rate=2)
0.25
```

Reply:
Cool, thanks.

I still wonder if we - since we support unit only - should also always support value only.

I.e.:

```python
>>> audmath.duration_in_seconds('s')
1.0
>>> audmath.duration_in_seconds('1.0')
1.0
>>> audmath.duration_in_seconds('1.0', sampling_rate=2)
0.5
```

Reply:
I changed it accordingly, removed the related `ValueError` from the docstring, and added another example to the docstring showing this behavior.

Reply:
But know we should use `sampling_rate` also if `duration` is not a string, I would argue.

Reply:
Yes, I think it makes sense. Then we avoid that the following two expressions have different return values:

```
>>> audmath.duration_in_seconds('1', sampling_rate=2)
0.5
>>> audmath.duration_in_seconds(1, sampling_rate=2)
1.0
```

There's only the question what we should do in the case the user provides a unit and sampling rate, e.g.:

```python
>>> audmath.duration_in_seconds('1s', sampling_rate=2)
1.0
```

At the moment we simply ignore the sampling rate. But I guess we can stick with this behavior.

Reply:
Yes, sounds good. Let me do some final checks and then we are good to go!

Reply:
I changed it accordingly, and now we have:

```python
>>> audmath.duration_in_seconds('1', sampling_rate=2)
0.5
>>> audmath.duration_in_seconds(1, sampling_rate=2)
0.5
>>> audmath.duration_in_seconds('1s', sampling_rate=2)
1.0
>>> audmath.duration_in_seconds(np.timedelta64(1, 's'), sampling_rate=2)
1.0
>>> audmath.duration_in_seconds(pd.to_timedelta(1, 's'), sampling_rate=2)
1.0
```